subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/WatchableHierarchies.java

```diff
@@ -135,11 +146,16 @@ private SnapshotHierarchy removeWatchedHierarchiesOverLimit(SnapshotHierarchy ro
                 result = invalidator.invalidate(locationToRemove.toString(), result);
             }
         }
+        rebuildWatchableFiles();
+        return result;
+    }
+
+    private void rebuildWatchableFiles() {
         this.watchableFiles = hierarchies.stream()
             .reduce(FileHierarchySet.empty(), FileHierarchySet::plus, Combiners.nonCombining());
```

Comment:
This side-effect-only method looks wrong to me. I'd much rather we returned a `FileHierarchySet` here and set it to `watchableFiles` wherever we call this.

Reply:
Done.